### PR TITLE
Implement custom helpers

### PR DESCRIPTION
**Pull request by juanazam**
Implements #22

### CODE REVIEW

**Comment by juanazam** [comment]
@san650 I have finished with the discussed suggestions, please check it out :D

File: test-support/page-object/custom-helper.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+          let scopedSelector = qualifySelector(parent.scope, selector);
+
+          let customHelperRes = userDefinedFunction(scopedSelector, options);
+
+          if ($.isPlainObject(customHelperRes)) {
+            return build($.extend(customHelperRes, { scope: scopedSelector }),
+                         key,
+                         parent);
+          } else if ($.isFunction(customHelperRes)) {
+            return customHelperRes(scopedSelector);
```
Comment (san650) [inline test-support/page-object/custom-helper.js]: I think we don't need to pass the scope as an argument here. We just need to invoke the function, right?

File: test-support/page-object/custom-helper.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        return function() {
+          // eval
+          let scopedSelector = qualifySelector(parent.scope, selector);
+
+          let customHelperRes = userDefinedFunction(scopedSelector, options);
+
+          if ($.isPlainObject(customHelperRes)) {
+            return build($.extend(customHelperRes, { scope: scopedSelector }),
+                         key,
+                         parent);
```
Comment (san650) [inline test-support/page-object/custom-helper.js]: I think the order of the parameters in the `extend` is not the best.

If the customHelperRes has a `scope: 'something'` attribute, we're overriding it with our `{ scope: scopedSelector }` and may not be the desired effect.

I think the order could be `$.extend({ scope: scopedSelector }, customHelperRes)`, this way the user can override the scope.

File: test-support/page-object/custom-helper.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  return function(selector, options) {
+    // def
+    return {
+      buildPageObjectAttribute: function(key, parent) {
+        // def
+        return function() {
+          // eval
+          let scopedSelector = qualifySelector(parent.scope, selector);
+
+          let customHelperRes = userDefinedFunction(scopedSelector, options);
```
Comment (san650) [inline test-support/page-object/custom-helper.js]: Maybe we can simplify the code by renaming some variables `s/customHelperRes/response` `s/scopedSelector/qualifiedSelector/`

Also, maybe we can improve readability by having only one return statement

```js
if ($.isPlainObject(response)) {
    processedResponse = build(...);
} else if ($.isFunction(response)) {
    processedResponse = response();
} else {
    processedResponse = response;
}

return processedResponse;
```

What do you think?

File: DOCUMENTATION.md [hunk 4]
```diff
@@ -615,6 +616,87 @@ var page = PO.build({
 
 Note that if the plain object doesn't have attributes defined, the object is returned as is.
 
+### `.customHelper`
+
+Defines a way in which custom components can be created using
+information of their surrounding context.
+
+It expects a function with the following signature:
```
Comment (san650) [inline DOCUMENTATION.md]: If you look at the other attributes, we add a _attribute signature_ section after the short description.

Something like...

```markdown
Attribute signature

` ``js
PO.customHelper(userFunction);
` ``

Where the provided user function has the following signature

` ``js
function(selector, options) {
  // user __code__ goes here
  return Value;
}
` ``

...

```

Or something on the line

File: DOCUMENTATION.md [hunk 5]
```diff
@@ -615,6 +616,87 @@ var page = PO.build({
...
+
+var disabled = PO.customHelper(function(selector, options) {
+  return $(selector).prop('disabled');
+});
+
+var page = PO.build({
+  myText: {
+    scope: '#myInput',
+    isAdminField: admin(),
+    isDisabled: disabled()
```
Comment (san650) [inline DOCUMENTATION.md]: Can we change `disabled` and `admin` function names to be `isDisabled` and `isAdmin`? So we follow the same convention as the rest of the project

**Comment by san650** [comment]
@juanazam This is great! Sorry I did a lot of comments but I think they're just small adjustments.

I'm going to test the functionality on a real project now and let you know how it goes :smile:

File: test-support/page-object/custom-helper.js (hunk 1 above)
Comment (juanazam) [inline test-support/page-object/custom-helper.js]: Following the spec we agreed on, we must pass the result of calling `qualifySelector(parent.scope, selector)` to get the parent's scope when the user defined function is called.

```
var input = customHelper(function(selector, options) {
  let hasError = $(selector).parent().hasClass('has-error');

  return {
    value: value(selector),
    hasError: function() {
      return hasError;
    }
  };
});

var page = PageObject.build({
  scope: 'form',
  userName: input('#userName')
});
```
In this case, the input function should receive `form #userName` as a parameter.

File: test-support/page-object/custom-helper.js (hunk 2 above)
Comment (juanazam) [inline test-support/page-object/custom-helper.js]: :+1:

File: test-support/page-object/custom-helper.js (hunk 3 above)
Comment (juanazam) [inline test-support/page-object/custom-helper.js]: :+1:

File: DOCUMENTATION.md (hunk 4 above)
Comment (juanazam) [inline DOCUMENTATION.md]: :+1:

**Comment by juanazam** [comment]
@san650 I addressed all issues except the discussion around if the selector should be passed as a parameter when the user defined function returns a function. :smile:

**Pull request merged**